File: src/app/servers.ts
import { Command } from 'commander';
import Session from '../session';
import { buildServer, parseFlagOptions } from '../validate';
import log from '../log';

const getServersCmd = new Command('get-servers')
    .addHelpText('before', 'Fetches all servers from the panel.')
    .option('--json', 'Send the response output as JSON.', true)
    .option('--yaml', 'Send the response output as YAML.', false)
    .option('--text', 'Send the response output as formatted text.', false)
    .option('-n, --no-prompt', 'Don\'t prompt for user response after the request.', false)
    .option('-s, --silent', 'Don\'t log request messages.', false)
    .option('-o, --output [file]', 'Writes the output to a file.')
    .option('--id <id>', 'The server ID to fetch.')
    .option('--uuid <uuid>', 'The UUID to query.')
    .option('--name <name>', 'The server name to query.')
    .option('--external <id>', 'The external server ID to query.')
    .option('--image <url>', 'The docker image URL to query.')
    .action(async (args: object) => {
        const options = parseFlagOptions(args);
        const session = new Session('application', options);

        const data = await session.handleRequest('GET', buildServer(args));
        const out = await session.handleClose(data, options);
        if (out) {
            if (!options.silent) log.success('request result:\n');
            console.log(out);
        }
    });

const suspendServerCmd = new Command('suspend')
    .addHelpText('before', 'Suspends a specified server.')
    .argument('<id>', 'The ID of the server.')
    .option('-s, --silent', 'Don\'t log request messages.', false)
    .action(async (id: string, args: object) => {
        const options = parseFlagOptions(args);
        const session = new Session('application', options);

        await session.handleRequest('POST', buildServer({ id, suspend: true }));
        if (!options.silent) log.success(`suspended server: ${id}`);
    });

    const unsuspendServerCmd = new Command('unsuspend')
    .addHelpText('before', 'Unsuspends a suspended server.')
    .argument('<id>', 'The ID of the server.')
    .option('-s, --silent', 'Don\'t log request messages.', false)
    .action(async (id: string, args: object) => {
        const options = parseFlagOptions(args);
        const session = new Session('application', options);

        await session.handleRequest('POST', buildServer({ id, suspend: true }));
        if (!options.silent) log.success(`unsuspended server: ${id}`);
    });

export default [
    getServersCmd,
    suspendServerCmd,
    unsuspendServerCmd
]

File: app/servers.go
package app

import (
	"errors"
	"fmt"
	"strings"

	"github.com/pteropackages/soar/config"
	"github.com/pteropackages/soar/http"
	"github.com/pteropackages/soar/util"
	"github.com/spf13/cobra"
)

var getServersCmd = &cobra.Command{
	Use:   "servers:get [--id id] [--external id] [--name name]\n\t[--desc info] [--uuid id] [--image name]",
	Short: "gets panel servers",
	Long:  getServersHelp,
	Run: func(cmd *cobra.Command, _ []string) {
		log.ApplyFlags(cmd.Flags())

		global, _ := cmd.Flags().GetBool("global")
		cfg, err := config.Get(global)
		if err != nil {
			config.HandleError(err, log)
			return
		}
		cfg.ApplyFlags(cmd.Flags())

		single, query, err := parseServerQuery(cmd)
		if err != nil {
			log.Error("command error:").WithError(err)
			return
		}

		ctx := http.New(cfg, &cfg.Application, log)
		req := ctx.Request("GET", "/api/application/servers"+query, nil)
		res, err := ctx.ExecuteWithFlags(req, cmd.Flags())
		if err != nil {
			log.WithError(err)
			return
		}

		var buf []byte

		if single {
			buf, err = http.HandleItemResponse(res, cfg)
		} else {
			buf, err = http.HandleDataResponse(res, cfg)
		}
		if err != nil {
			log.WithError(err)
			return
		}

		log.LineB(buf)
	},
}

func parseServerQuery(cmd *cobra.Command) (bool, string, error) {
	var query strings.Builder
	var params []string
	single := false
	flags := cmd.Flags()

	if id, _ := flags.GetInt("id"); id != 0 {
		single = true
		query.WriteString(fmt.Sprintf("/%d", id))
	}

	if ext, _ := flags.GetString("external"); ext != "" {
		if query.Len() != 0 {
			return false, "", errors.New("id and external flags specified; pick one")
		}

		single = true
		query.WriteString("/external/" + ext)
	}

	if value, _ := flags.GetString("name"); value != "" {
		params = append(params, "filter[name]="+value)
	}

	if value, _ := flags.GetString("desc"); value != "" {
		params = append(params, "filter[description]="+value)
	}

	if value, _ := flags.GetString("uuid"); value != "" {
		params = append(params, "filter[uuid]="+value)
	}

	if value, _ := flags.GetString("image"); value != "" {
		params = append(params, "filter[image]="+value)
	}

	if len(params) > 0 {
		query.WriteString("?" + params[0])

		if len(params) > 1 {
			for _, p := range params {
				query.WriteString("&" + p)
			}
		}
	}

	return single, query.String(), nil
}

var suspendServerCmd = &cobra.Command{
	Use:   "servers:suspend id",
	Short: "suspends a server",
	Long:  "Suspends a server on the panel by its ID.",
	Run: func(cmd *cobra.Command, args []string) {
		log.ApplyFlags(cmd.Flags())
		if err := util.RequireArgs(args, []string{"id"}); err != nil {
			log.WithError(err)
			return
		}

		global, _ := cmd.Flags().GetBool("global")
		cfg, err := config.Get(global)
		if err != nil {
			config.HandleError(err, log)
			return
		}
		cfg.ApplyFlags(cmd.Flags())

		ctx := http.New(cfg, &cfg.Application, log)
		req := ctx.Request("POST", "/api/application/servers/"+args[0]+"/suspend", nil)
		if _, err = ctx.Execute(req); err != nil {
			log.WithError(err)
		}
	},
}

var unsuspendServerCmd = &cobra.Command{
	Use:   "servers:unsuspend id",
	Short: "unsuspends a server",
	Long:  "Unsuspends a server on the panel by its ID.",
	Run: func(cmd *cobra.Command, args []string) {
		log.ApplyFlags(cmd.Flags())
		if err := util.RequireArgs(args, []string{"id"}); err != nil {
			log.WithError(err)
			return
		}

		global, _ := cmd.Flags().GetBool("global")
		cfg, err := config.Get(global)
		if err != nil {
			config.HandleError(err, log)
			return
		}
		cfg.ApplyFlags(cmd.Flags())

		ctx := http.New(cfg, &cfg.Application, log)
		req := ctx.Request("POST", "/api/application/servers/"+args[0]+"/unsuspend", nil)
		if _, err = ctx.Execute(req); err != nil {
			log.WithError(err)
		}
	},
}

var reinstallServerCmd = &cobra.Command{
	Use:   "servers:reinstall id",
	Short: "reinstalls a server",
	Long:  "Triggers the reinstall process for a server by its ID.",
	Run: func(cmd *cobra.Command, args []string) {
		log.ApplyFlags(cmd.Flags())
		if err := util.RequireArgs(args, []string{"id"}); err != nil {
			log.WithError(err)
			return
		}

		global, _ := cmd.Flags().GetBool("global")
		cfg, err := config.Get(global)
		if err != nil {
			config.HandleError(err, log)
			return
		}
		cfg.ApplyFlags(cmd.Flags())

		ctx := http.New(cfg, &cfg.Application, log)
		req := ctx.Request("POST", "/api/application/servers/"+args[0]+"/reinstall", nil)
		if _, err = ctx.Execute(req); err != nil {
			log.WithError(err)
		}
	},
}

var deleteServerCmd = &cobra.Command{
	Use:   "servers:delete id [--force]",
	Short: "deletes a server",
	Long:  "Deletes a server on the panel by its ID (supports the --force flag).",
	Run: func(cmd *cobra.Command, args []string) {
		log.ApplyFlags(cmd.Flags())
		if err := util.RequireArgs(args, []string{"id"}); err != nil {
			log.WithError(err)
			return
		}

		global, _ := cmd.Flags().GetBool("global")
		cfg, err := config.Get(global)
		if err != nil {
			config.HandleError(err, log)
			return
		}
		cfg.ApplyFlags(cmd.Flags())

		path := "/api/application/servers/" + args[0]
		if force, _ := cmd.Flags().GetBool("force"); force {
			path += "/force"
		}

		ctx := http.New(cfg, &cfg.Application, log)
		req := ctx.Request("DELETE", path, nil)
		if _, err = ctx.Execute(req); err != nil {
			log.WithError(err)
		}
	},
}
